File: leetcode/103. Binary Tree Zigzag Level Order Traversal.cpp
#include <iostream>
#include <unordered_set>
#include <algorithm>
#include <vector>
#include <string>
#include <numeric>
#include <queue>
#include <stack>
using namespace std;
/**
 * Definition for a binary tree node.
 *  */
struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  };

class Solution {
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
        vector<vector<int> >res;
        if (!root) return res;
        stack<TreeNode*> s1;
        stack<TreeNode*> s2;
        s1.push(root);
        vector<int> out;
        while (!s1.empty() || !s2.empty()) {
            while (!s1.empty()) {
                TreeNode *cur = s1.top();
                s1.pop();
                out.push_back(cur->val);
                if (cur->left) s2.push(cur->left);
                if (cur->right) s2.push(cur->right);
            } 
            if (!out.empty()) res.push_back(out);
            out.clear();
            while (!s2.empty()) {
                TreeNode *cur = s2.top();
                s2.pop();
                out.push_back(cur->val);
                if (cur->right) s1.push(cur->right);
                if (cur->left) s1.push(cur->left);
            }
            if (!out.empty()) res.push_back(out);
            out.clear();
        }
        return res;
    }
};

File: leetcode/103. Binary Tree Zigzag Level Order Traversal.py
# Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).

# For example:
# Given binary tree [3,9,20,null,null,15,7],
#     3
#    / \
#   9  20
#     /  \
#    15   7
# return its zigzag level order traversal as:
# [
#   [3],
#   [20,9],
#   [15,7]
# ]

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
import queue
class Solution:
    def zigzagLevelOrder(self,root:TreeNode):
        ## 使用两个队列 
        if root == None:
            return []
        q = []
        q.append(root)
        res = []
        ans = []
        flag = 1
        while len(q)!=0:
            n = len(q)
            new_q = []
            for i in range(n):
                node = q[i]
                ans.append(node.val)
                if node.left!=None:
                    new_q.append(node.left)
                if node.right!=None:
                    new_q.append(node.right)
            if flag == -1:
                ans = ans[::-1]
            flag = -flag
            res.append(ans)
            ans=[]
            q = new_q
        return res 
    
        